### ISSUE TITLE
Support for additional dimensions in Dense layer

### ISSUE THREAD

**Issue by harryfei** [FluxML/Flux.jl#282]
Currently, the `Dense` layer convert a (in_features, N) matrix into a (out_features, N) matrix. Shall we extend this api to support additional dimensions? 

By supporting additional dimensions, the interface of `Dense` layer would be like this.
* Input Shape: (in_features, d1, d2, d3, ..., N)
* Ouput Shape: (out_features, d2, d2, d3,..., N)

References:
* [pytorch.Linear](https://pytorch.org/docs/stable/nn.html#linear)
* [keras.Dense](https://keras.io/layers/core/#dense)

**Comment by ninjin** [comment]
Yes, I think that sounds like a very elegant API that makes expressing something very common easy to read and less verbose. It should also be fairly easy to implement, would you be happy to submit a PR?

**Comment by fangyuanziti** [comment]
Of course, I will submit a PR for this. :smile:

**Comment by MikeInnes** [comment]
I take it the inner dimensions are effectively treated as extra batch dimensions?

It seems to me like you could go either way here, e.g. they could equally belong to the input features. But if this the more common / standard case and people are generally on board, I'm ok with it.

**Comment by fangyuanziti** [comment]
Actually, I have no idea about whether this the more common/standard case or not. Howerver, I found this inteface is very useful when I try modeling neural networks to process audio.

Besides, I don't think we can use input_features get the same thing as  additional dimensions.

If we use input_feautres only:

* Input Shape: (in_features * d1 * d2 * d3 * ..., N)
* Ouput Shape: (out_features * d1 * d2 *d3,..., N)
Finally, we get a dense layer with parameter size of (out_features * d1 * d2 * d3 * ..., in_features * d1 * d2 * d3 * ...).

If we can use additional dimensions.
* Input Shape: (in_features, d1, d2, d3, ..., N)
* Ouput Shape: (out_features, d1, d2, d3,..., N)
Finally, we get a dense layer whit parameter size of (out_features, in_features). 

The dense layers in the two situations are different. So I suggest to add addtional dimensions in `Dense` layer.

**Comment by MikeInnes** [comment]
Actually, they are not necessarily exclusive, as long as one is explicit. For example you could support N-dimensional features with `Dense((5, 3), (2, 7))`. Then this accepts a `(5, 3, ...)` array and has a `15x14` weight matrix.

**Comment by fangyuanziti** [comment]
@MikeInnes Oh, Thanks for your explanation. I got your idea. Howerver, I'm afraid it doesn't meet my case.

The differenence between input and ouput's shapes is just `features` axis ( in additoinal dimenisons). 
eg. 
  * [40, 2, 2, 1000] -> [20, 2, 2, 1000]. The expected shape of  weight matrix is `20 x 40`, not `80 x 160`.

**Comment by ChrisRackauckas** [comment]
Internal layers could make use of EllipsisNotation.jl so that way it easily handles any size.

**Comment by darsnack** [comment]
This is closed by #1405
